phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java

```diff
@@ -178,8 +178,11 @@ private static void throwIfScanOutOfRegion(Scan scan, Region region) throws DoNo
             byte[] expectedUpperRegionKey =
                     scan.getAttribute(EXPECTED_UPPER_REGION_KEY) == null ? scan.getStopRow() : scan
                             .getAttribute(EXPECTED_UPPER_REGION_KEY);
-            isStaleRegionBoundaries = expectedUpperRegionKey != null &&
-                    Bytes.compareTo(upperExclusiveRegionKey, expectedUpperRegionKey) != 0;
+
+            byte[] actualStartRow = scan.getAttribute(SCAN_ACTUAL_START_ROW);
+            isStaleRegionBoundaries = (expectedUpperRegionKey != null &&
+                    Bytes.compareTo(upperExclusiveRegionKey, expectedUpperRegionKey) != 0) || 
+                    (actualStartRow != null && Bytes.compareTo(actualStartRow, lowerInclusiveRegionKey) < 0);
         } else {
             isStaleRegionBoundaries = Bytes.compareTo(lowerInclusiveScanKey, lowerInclusiveRegionKey) < 0 ||
```

Comment:
Why is this logic different for the non-local index case? I guess the only thing that should change is how we extract the start and end keys of the scan, right? For ex, in the case of a local index scan, these would be passed as scan attributes and for normal scans they are just the startRow and endRow of the scan, or am I not understanding this correctly?

Reply:
This is because for local indexes we cannot (ever) proceed when the region was split, we have to restart, since we cannot exactly know where the index keys were split to, they sort differently than the main table's keys.
It's not enough that the new scan is fully "contained" in the new region. Is that end "old" scan boundaries are contained on the lower side and exact on the upper side.

Reply:
Makes sense. Thanks for the explanation.